lib/AST/ModuleDependencies.cpp

```diff
@@ -332,24 +332,137 @@ void ModuleDependenciesCache::recordDependencies(
+    return map[moduleName].end()-1;
   } else {
     map.insert({moduleName, ModuleDependenciesVector{std::move(dependencies)}});
     AllModules.push_back({moduleName.str(), kind});
+    return &(map[moduleName].front());
   }
 }
 
-void ModuleDependenciesCache::updateDependencies(
+const ModuleDependencies* GlobalModuleDependenciesCache::updateDependencies(
     ModuleDependencyID moduleID, ModuleDependencies dependencies) {
   auto &map = getDependenciesMap(moduleID.second);
   auto known = map.find(moduleID.first);
   assert(known != map.end() && "Not yet added to map");
-  assert(known->second.size() == 1 && "Cannot update dependency with multiple candidates.");
+  assert(known->second.size() == 1 &&
+         "Cannot update dependency with multiple candidates.");
   known->second[0] = std::move(dependencies);
+  return &(known->second[0]);
+}
+
+llvm::StringMap<const ModuleDependencies*> &
+ModuleDependenciesCache::getDependencyReferencesMap(ModuleDependenciesKind kind) {
+  switch (kind) {
+  case ModuleDependenciesKind::SwiftTextual:
+    return SwiftTextualModuleDependencies;
+  case ModuleDependenciesKind::SwiftBinary:
+    return SwiftBinaryModuleDependencies;
+  case ModuleDependenciesKind::SwiftPlaceholder:
+    return SwiftPlaceholderModuleDependencies;
+  case ModuleDependenciesKind::Clang:
+    return ClangModuleDependencies;
+  }
+  llvm_unreachable("invalid dependency kind");
+}
+
+const llvm::StringMap<const ModuleDependencies*> &
+ModuleDependenciesCache::getDependencyReferencesMap(ModuleDependenciesKind kind) const {
+  switch (kind) {
+  case ModuleDependenciesKind::SwiftTextual:
+    return SwiftTextualModuleDependencies;
+  case ModuleDependenciesKind::SwiftBinary:
+    return SwiftBinaryModuleDependencies;
+  case ModuleDependenciesKind::SwiftPlaceholder:
+    return SwiftPlaceholderModuleDependencies;
+  case ModuleDependenciesKind::Clang:
+    return ClangModuleDependencies;
+  }
+  llvm_unreachable("invalid dependency kind");
+}
+
+bool ModuleDependenciesCache::hasDependencies(
+    StringRef moduleName, Optional<ModuleDependenciesKind> kind) const {
+  if (!kind) {
+    return hasDependencies(moduleName, ModuleDependenciesKind::SwiftTextual) ||
+           hasDependencies(moduleName, ModuleDependenciesKind::SwiftBinary) ||
+           hasDependencies(moduleName,
+                           ModuleDependenciesKind::SwiftPlaceholder) ||
+           hasDependencies(moduleName, ModuleDependenciesKind::Clang);
+  }
+  const auto &map = getDependencyReferencesMap(*kind);
+  return map.count(moduleName) > 0;
+}
+
+bool ModuleDependenciesCache::hasDependencies(
+    StringRef moduleName, ModuleLookupSpecifics details) const {
+  // 1. Query the local scan results
+  // 2. If no module found, query the global cache using the module details
+  // lookup
+  if (hasDependencies(moduleName, details.kind))
+    return true;
+  else
+    return globalCache.hasDependencies(moduleName, details);
```

Comment:
Can we delegate the logics in `hasDependencies` to `findDependencies`?

Reply:
Good idea. Now only `findDependencies` goes looking for stuff and `hasDependencies` just checks if anything was found.